### DISCUSSION TITLE
Shelly EM does not show negative values

### DISCUSSION THREAD

**Discussion by pauloamexio** [arendst/Tasmota#12861] [Ideas]
Hello,

I have two Shelly's to monitor the energy. One Shelly 1PM that monitors the inbound energy from the solar panels and one Shelly EM to monitor the inbound/outbound energy that comes from distributor.

My issue is the Shelly EM is always giving an positive power value when the energy is inbound or outbound from the distributor, which does not make any sense. Do you know how may I fix this issue?

Bellow I send some graph images from Shelly EM and Shelly 1PM where at value 18:08:00 the value should be negative in the Shelly EM graph.

![image](https://user-images.githubusercontent.com/87972768/129008508-7fbab9da-8bfe-40c9-85d6-284d7b199be5.png)
![image](https://user-images.githubusercontent.com/87972768/129008575-49d42926-7364-4b6c-935f-69dab92e8255.png)

Shelly 1PM: https://shelly.cloud/knowledge-base/devices/shelly-1pm/
Shelly EM: https://shelly.cloud/knowledge-base/devices/shelly-em/

Shelly EM chip configuration: https://www.analog.com/en/products/ade7953.html
Shelly EM ADE7953 datasheet: https://www.analog.com/media/en/technical-documentation/data-sheets/ADE7953.pdf

Tasmota ADE7953 source code: https://github.com/arendst/Tasmota/blob/development/tasmota/xnrg_07_ade7953.ino

Best regards,
Paulo

**Comment by Jason2866** [comment]
If i am not wrong, with my quick look at the driver, it is not supported from the driver.
Driver would need an enhancement for reading the sign indication register.
From the datasheet:
```
SIGN INDICATION The ADE7953 includes four sign indication bits that indicate the polarity of the active power on Current Channel A (APSIGN_A), the active power on Current Channel B (APSIGN_B), the reactive power on Current Channel A (VARSIGN_A), and the reactive power on Current Channel B (VARSIGN_B). These bits are located in the ACCMODE register (Address 0x201 and Address 0x301). All four bits are unlatched and read only. A low reading (0) on any of these bits indicates that the corresponding power reading is positive; a high reading (1) indicates that the corresponding power reading is negative. These bits are enabled by default and are disabled in the corresponding no-load condition.
```

**Reply by pauloamexio** [reply]
So it's not possible to check with tasmota if it's an high or an low reading?
It's strange because before I had solar panels,  I put the shelly em clamp  inverted and the Power consumption was giving negative values with no issue (using tasmota).

**Comment by pauloamexio** [comment]
Hello, 

I did modification in the driver as @Jason2866 said in this pull request: https://github.com/arendst/Tasmota/pull/12874

I read the ACCMODE register according the data sheet table on page 65:

![image](https://user-images.githubusercontent.com/87972768/129185550-07966dbf-3528-4a48-b13a-2ba3e7dd3dc7.png)

Now the appears as negative in the active power:

![image](https://user-images.githubusercontent.com/87972768/129185660-709cdaaa-b05b-49db-817b-873d4867077c.png)

The graphs in grafana:

![image](https://user-images.githubusercontent.com/87972768/129185826-b7c2a239-29f4-4ef8-b656-bfbeb7e00c2c.png)

![image](https://user-images.githubusercontent.com/87972768/129185899-c2ec2d4f-004d-4bc4-a7e5-b3e2886b31eb.png)

This modification looks ok for you guys? May I create a new issue from this discussion?

**Reply by Jason2866** [reply]
Thx for your PR!

**Reply by pauloamexio** [reply]
You're welcome :)
